### PR TITLE
Add support for Protobuf v22+

### LINKED ISSUE #10133: Add support for Protobuf v22+
### Description

There are some breaking changes in the 22 release and we need to investigate and implement any required changes to support v22+.

### PR DESCRIPTION
**Pull request by tigrux**
This change adds compatibility with newer versions of Protobuf.
The actual version of Protobuf is not being updated.
Protobuf v27.2 was used for testing.
Protobuf v21.8 is still supported.

Fixes #10133

### CODE REVIEW

**Comment by tigrux** [comment]
Hello @majetideepak, @assignUser, @pedroerp.
I have been experimenting with support for newer versions of Protobuf (v27.1) in Velox.
I had started this effort while I was trying to add cmake modules for GCS, but we agreed to propose the changes in a different PR.

File: CMake/resolve_dependency_modules/protobuf.cmake [hunk 1]
```diff
@@ -13,15 +13,15 @@
 # limitations under the License.
 include_guard(GLOBAL)
 
-set(VELOX_PROTOBUF_BUILD_VERSION 21.8)
+set_source(absl)
+resolve_dependency(absl 20240116 EXACT)
+
+set(VELOX_PROTOBUF_BUILD_VERSION 27.1)
```
Comment (pedroerp) [inline CMake/resolve_dependency_modules/protobuf.cmake:19]: Does that mean we will now require a newer version of protobuf - meaning 21.8 will now be unsupported? Do we know which version do most users still use? For example, which protobuf version ships with current popular distros like Ubuntu, Fedora, Centos, etc?

Comment (tigrux) [reply CMake/resolve_dependency_modules/protobuf.cmake:19]: Keep in mind that it is more of a concept so it won't be submitted as final.
The first commit adds backwards and forward compatibility, and I think it can be submitted as a different PR.
I added the second commit precisely to test that both protobuf versions are supported, and it should be submitted as another PR as needed, or maybe pick the versions from a manifest, I don't know what is best.

**Comment by tigrux** [comment]
Hello again @majetideepak, @assignUser, @pedroerp, @kgpai.
I was able to fix the issues of the previous attempts.
Please review, and best regards.

File: CMake/resolve_dependency_modules/protobuf.cmake (hunk 1 above)
Comment (tigrux) [reply CMake/resolve_dependency_modules/protobuf.cmake]: In the last PR, now in its final form, we only have to modify the .cmake modules and the setup-centos, setup-adapters files. I would like to centralize the versioning of the dependencies but that may be a future improvement.

File: velox/substrait/tests/JsonToProtoConverter.cpp [hunk 2]
```diff
@@ -33,5 +33,10 @@ void JsonToProtoConverter::readFromFile(
   VELOX_CHECK(
       status.ok(),
       "Failed to parse Substrait JSON: {}",
-      status.message().ToString());
+#if PROTOBUF_VERSION < 4022000
```
Comment (majetideepak) [inline velox/substrait/tests/JsonToProtoConverter.cpp:36]: Can we use `std::string_view(status.message.data(), static_cast<size_type>(status.message.size()))` and avoid this ifdef? The above API is available in both `absl::string_view` and older protobuf `stringpiece`.

**Review by majetideepak — COMMENTED** [review]
@tigrux one code comment. The CMake changes look good to me.
But I want @assignUser to make a pass as well.

File: CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake [hunk 3]
```diff
@@ -13,11 +13,8 @@
 # limitations under the License.
 include_guard(GLOBAL)
 
-set_source(absl)
-resolve_dependency(absl 20240116 EXACT)
-
 set_source(gRPC)
-resolve_dependency(gRPC 1.48.1 EXACT)
+resolve_dependency(gRPC CONFIG REQUIRED)
```
Comment (assignUser) [inline CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake:17]: You don't have to enforce `CONFIG` mode as it's the default, it will only look for a `Find*.cmake` if no config is found. 

If there is no specific reason why this is necessary I would prefer to keep the option open for people to supply find modules if necessary for what ever reason (otherwise they would need to patch this out... like we have done for e.g. folly^^). This goes for the other uses of `CONFIG` as well of course.

File: CMakeLists.txt [hunk 4]
```diff
@@ -442,7 +442,7 @@ if(${VELOX_BUILD_MINIMAL_WITH_DWIO}
 
   # Locate or build protobuf.
   set_source(Protobuf)
-  resolve_dependency(Protobuf 3.21.7...<4)
+  resolve_dependency(Protobuf CONFIG REQUIRED)
```
Comment (assignUser) [inline CMakeLists.txt:445]: Do we really want to drop the lower bound too?

File: CMakeLists.txt [hunk 5]
```diff
@@ -470,7 +470,7 @@ endif()
 
 if(VELOX_ENABLE_GCS)
   set_source(google_cloud_cpp_storage)
-  resolve_dependency(google_cloud_cpp_storage 2.22.0 EXACT)
+  resolve_dependency(google_cloud_cpp_storage CONFIG REQUIRED)
```
Comment (assignUser) [inline CMakeLists.txt:473]: Do we really want to drop the lower bound?

File: velox/dwio/dwrf/proto/CMakeLists.txt [hunk 6]
```diff
@@ -30,12 +30,15 @@ endforeach()
 set(PROTO_OUTPUT_FILES ${PROTO_HDRS} ${PROTO_SRCS})
 set_source_files_properties(${PROTO_OUTPUT_FILES} PROPERTIES GENERATED TRUE)
 
+foreach(PROTO_PATH ${CMAKE_SOURCE_DIR} ${Protobuf_INCLUDE_DIRS})
+  list(APPEND PROTO_PATH_ARGS --proto_path=${PROTO_PATH})
+endforeach()
```
Comment (assignUser) [inline velox/dwio/dwrf/proto/CMakeLists.txt:35]: Why is this needed vs. the previous version where it's hardcoded in the  `COMMAND`?  Is `PROTO_PATH_ARGS` generated by the protobuf config or something? Could you add a comment?

File: CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake (hunk 3 above)
Comment (assignUser) [reply CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake:17]: Also dropping the minimum?

**Review by assignUser — APPROVED** [review]
Some minor comments but looks good overall, thanks!

File: CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake (hunk 3 above)
Comment (tigrux) [reply CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake:17]: Without the CONFIG, I was getting the weird errors about missing absl symbols.

File: velox/substrait/tests/JsonToProtoConverter.cpp (hunk 2 above)
Comment (tigrux) [reply velox/substrait/tests/JsonToProtoConverter.cpp]: Fixed!

File: CMakeLists.txt (hunk 4 above)
Comment (tigrux) [reply CMakeLists.txt:445]: Yes, when protobuf is preinstalled, it reports version 5.27.2.
However, when it is used as cmake module it reports version 27.2.
The same happen happens with the current version (21.8 vs 3.21.8).
It also complicates versioning the dependencies, because we need to keep track of every cmake file that refers to a version. I may put back the explicit versions if you disagree.

File: CMakeLists.txt (hunk 5 above)
Comment (tigrux) [reply CMakeLists.txt:473]: See my comment above about protobuf.

File: CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake (hunk 3 above)
Comment (tigrux) [reply CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake:17]: See my comment about protobuf.

File: velox/dwio/dwrf/proto/CMakeLists.txt (hunk 6 above)
Comment (tigrux) [reply velox/dwio/dwrf/proto/CMakeLists.txt:35]: For some weird reason, with the previous line...
  --proto_path ${Protobuf_INCLUDE_DIRS} --cpp_out ${PROJECT_SOURCE_DIR}
...the MacOS builds failed with "--proto_path without argument".
This because the variable Protobuf_INCLUDE_DIRS is empty when Protobuf is installed in the system.
So I was forced to make sure that --proto_path is passed an argument.

File: CMakeLists.txt (hunk 4 above)
Comment (majetideepak) [reply CMakeLists.txt:445]: We should keep the minimum `3.21.7 (21.7)`. I wonder if there is a way to normalize the version numbers.

File: velox/dwio/dwrf/proto/CMakeLists.txt (hunk 6 above)
Comment (majetideepak) [reply velox/dwio/dwrf/proto/CMakeLists.txt:35]: Let's add a comment for future reference.

Comment (tigrux) [reply velox/dwio/dwrf/proto/CMakeLists.txt:36]: Agree. I added comments.

File: CMakeLists.txt (hunk 5 above)
Comment (assignUser) [reply CMakeLists.txt:473]: I don't understand what the protobuf version has to do with the other versions? There must still be a minimum version we need? If there are pairings between versions  of the dependencies we need we should at the minimum document them but ideally detect and properly check them.

File: CMakeLists.txt (hunk 4 above)
Comment (assignUser) [reply CMakeLists.txt:445]: The official findprotobuf.cmake has pretty extensive logic for the versions, maybe that can handle the different version schemas? (E.g. add `MODULE`) https://github.com/Kitware/CMake/blob/master/Modules%2FFindProtobuf.cmake#L590

Comment (tigrux) [reply CMakeLists.txt:445]: Agree. I am investigating this.

Comment (assignUser) [reply CMakeLists.txt:445]: Also we only care about the pre installed version here as our bundled version is fixed and not checked.

Comment (tigrux) [reply CMakeLists.txt]: Fixed. I had not realized that by using CONFIG in the find_package, we always get the version that has the ABI as prefix, 3.21.7 in this case. I also removed the <4 because the newest protobuf is now in ABI version 5.

File: CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake (hunk 3 above)
Comment (majetideepak) [reply CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake:17]: Should we keep the minimum version here as well?

Comment (tigrux) [reply CMake/resolve_dependency_modules/google_cloud_cpp_storage.cmake]: I just put back the minimum version, to advertise the version that we know that works. Thanks.

**Review by majetideepak — APPROVED** [review]
Thanks, @tigrux
